README.md

```diff
@@ -157,6 +188,300 @@ the next request either as a hidden form field or via JavaScript async request.
+
+#### Configure CORS module
+
+- A successful responses to a CORS preflight requests **must** contain appropriate CORS headers.
+- Responses to an actual requests **must** contain appropriate CORS headers.
+- Value of the CORS header `Access-Control-Allow-Origin` **must** contains origin from the predefined list.
+
+```
+// assuming frontend origin is https://example.com and backend origin is https://api.example.com
+Access-Control-Allow-Origin: https://example.com
+```
+
+#### Configure middlewares stack
+
+Add `CsrfHeaderMiddleware` to the main middleware stack:
+
+```php
+$middlewareDispatcher = $injector->make(MiddlewareDispatcher::class);
+$middlewareDispatcher = $middlewareDispatcher->withMiddlewares([
+    ErrorCatcher::class,
+    CsrfHeaderMiddleware::class, // <-- add this
+    Router::class,
+]);
+```
+
+or to the routes that must be protected:
+
+```php
+$collector = $container->get(RouteCollectorInterface::class);
+$collector->addGroup(
+    Group::create('/api')
+        ->middleware(CsrfHeaderMiddleware::class) // <-- add this
+        ->routes($routes)
+);
+```
+
+#### Configure frontend requests
+
+On the frontend add to the `GET`, `HEAD`, `POST` requests a custom header defined in the `CsrfHeaderMiddleware` with an empty or random value.
+
+```js
+let response = fetch('https://api.example.com/whoami', {
+  headers: {
+    "X-CSRF-HEADER": crypto.randomUUID()
+  }
+});
+```
+
+### Protecting requests passed from any origin
+
+In this scenario:
+- AJAX/SPA frontend and API backend have different origins.
+- Allow cross origin requests to the API server from any origin.
+- All requests are considered unsafe and **must** be protected against CSRF with CSRF-token.
+
+#### Configure CORS module
+
+- A successful responses to a CORS preflight requests **must** contain appropriate CORS headers.
+- Responses to an actual requests **must** contain appropriate CORS headers.
+- The CORS header `Access-Control-Allow-Origin` has the same value as `Origin` header in the request.
+
+```
+$frontendOrigin = $request->getOrigin();
+
+Access-Control-Allow-Origin: $frontendOrigin
+```
+
+#### Configure middlewares stack
+
+By default, `CsrfMiddleware` considers `GET`, `HEAD`, `OPTIONS` methods as safe operations and doesn't perform CSRF validation.
+In JavaScript-based apps, requests are made programmatically; therefore, to increase application protection, the only `OPTIONS` method can be considered safe and need not be appended with a CSRF token header.
+
+Configure `CsrfMiddleware` safe methods:
+
+```php
+$csrfMiddleware = $container->get(CsrfMiddleware::class);
+$csrfMiddleware = $csrfMiddleware->withSafeMethods([Method::OPTIONS]);
+```
+
+Add `CsrfMiddleware` to the main middleware stack:
+
+```php
+$middlewareDispatcher = $injector->make(MiddlewareDispatcher::class);
+$middlewareDispatcher = $middlewareDispatcher->withMiddlewares([
+    ErrorCatcher::class,
+    SessionMiddleware::class,
+    CsrfMiddleware::class, // <-- add this
```

Comment:
Should it use an instance that has OPTIONS set?

Reply:
I've added example for the di config as a workaround.